### PR TITLE
Fix int/uint interoperation to conform to de-facto expectations of other clients

### LINKED ISSUE #134: question: on uint vs. int decoding.
I started getting some of these errors on decode:

```
attempted to decode type "uint" with method for "int"
```

The culprit was another msgpack implementation

```
func (e *msgpackEncDriver) EncodeInt(i int64) {
    if i >= 0 {
        e.EncodeUint(uint64(i))
```

I found a few other libraries doing this trick.  Any suggestions here on how i can morph your library to read them?

update:

I'll spare you reading lua code, but  
https://github.com/fperrad/lua-MessagePack/

also defaults to unsigned.

```
set_integer'unsigned'
```

update, and one version of python, basically the pattern is "if non-negative, write as unsigned, if negative write as signed"

```

# You may notice struct.pack("B", obj) instead of the simpler chr(obj) in the
# code below. This is to allow for seamless Python 2 and 3 compatibility, as
# chr(obj) has a str return type instead of bytes in Python 3, and
# struct.pack(...) has the right return type in both versions.

def _pack_integer(obj, fp):
    if obj < 0:
        if obj >= -32:
            fp.write(struct.pack("b", obj))
        elif obj >= -2**(8-1):
            fp.write(b"\xd0" + struct.pack("b", obj))
        elif obj >= -2**(16-1):
            fp.write(b"\xd1" + struct.pack(">h", obj))
        elif obj >= -2**(32-1):
            fp.write(b"\xd2" + struct.pack(">i", obj))
        elif obj >= -2**(64-1):
            fp.write(b"\xd3" + struct.pack(">q", obj))
        else:
            raise UnsupportedTypeException("huge signed int")
    else:
        if obj <= 127:
            fp.write(struct.pack("B", obj))
        elif obj <= 2**8-1:
            fp.write(b"\xcc" + struct.pack("B", obj))
        elif obj <= 2**16-1:
            fp.write(b"\xcd" + struct.pack(">H", obj))
        elif obj <= 2**32-1:
            fp.write(b"\xce" + struct.pack(">I", obj))
        elif obj <= 2**64-1:
            fp.write(b"\xcf" + struct.pack(">Q", obj))
        else:
            raise UnsupportedTypeException("huge unsigned int")
```

### PR DESCRIPTION
**Pull request by shabbyrobe**
Been running this in production for a while now. This fixes #134 to meet the behaviour of other clients, which encode non-negative integers as unsigned regardless of whether the originating type is signed.

It raises an error on underflow or overflow and is tested pretty thoroughly (though I'm always happy to add more).

---
This change is [Reviewable](https://reviewable.io/reviews/tinylib/msgp/215)

### CODE REVIEW

**Comment by dchenk** [comment]
I like this, but I don't see why ReadUint64 needs to be extended to signed integers. It's unlikely that feature will ever come in useful to anyone.

**Comment by shabbyrobe** [comment]
If it doesn't hurt, and it works, why not? The spec is totally unclear about this one way or the other. What's to stop some client from implementing it the other way, and who would we be to say they were wrong?

File: msgp/read_bytes_test.go [hunk 1]
```diff
@@ -222,53 +225,211 @@ func BenchmarkReadBoolBytes(b *testing.B) {
...
+		return false
+	}
+
+	vs := []struct {
+		v        interface{}
+		rdBits   int
+		failBits int
+		errCheck func(err error, failBits int) bool
+	}{
+		{math.MaxInt64, i32, 32, overflowErr},
```
Comment (philhofer) [inline msgp/read_bytes_test.go]: these need explicit casts in order to make the 386 build pass

**Comment by shabbyrobe** [comment]
Oops, yep, sorry about that. I've fixed it up and it's passing on my machine with `GOARCH=386`, hopefully Travis doesn't spit chips!

**Pull request merged**